shared/ssl.sh

```diff
@@ -0,0 +1,122 @@
+CSRDIR="${SSLDIR}/certificate_requests"
+CERTDIR="${SSLDIR}/certs"
+mkdir -p "${SSLDIR}" "${PUBKEYDIR}" "${PRIVKEYDIR}" "${CSRDIR}" "${CERTDIR}"
+PUBKEYFILE="${PUBKEYDIR}/${CERTNAME}.pem"
+PRIVKEYFILE="${PRIVKEYDIR}/${CERTNAME}.pem"
+CSRFILE="${CSRDIR}/${CERTNAME}.pem"
+CERTFILE="${CERTDIR}/${CERTNAME}.pem"
+CACERTFILE="${CERTDIR}/ca.pem"
+
+CA="https://${PUPPETSERVER_HOSTNAME}:8140/puppet-ca/v1"
+CERTSUBJECT="/CN=${CERTNAME}"
+CERTHEADER="-----BEGIN CERTIFICATE-----"
+CURLFLAGS="--silent --show-error --cacert ${CACERTFILE}"
+
+### Print configuration for troubleshooting
+msg "Getting signed certificate for '${CERTNAME}' (${CERTSUBJECT})"
+msg "Using Puppet CA at '${CA}'"
+msg "Saving SSL files in '${SSLDIR}'"
+msg "Waiting up to ${WAITFORCERT} seconds for certificate to be signed"
+
+### Get the CA certificate for use with subsequent requests
+### Fail-fast if curl errors or the CA certificate can't be parsed
+curl --insecure --silent --show-error --output "${CACERTFILE}" "${CA}/certificate/ca"
+if [ $? -ne 0 ]; then
+    error "cannot reach CA host '${PUPPETSERVER_HOSTNAME}'"
+elif ! openssl x509 -subject -issuer -noout -in "${CACERTFILE}"; then
+    error "invalid CA certificate"
+fi
+
+### Check the CA does not already have a signed certificate for this host
+output="$(curl ${CURLFLAGS} "${CA}/certificate/${CERTNAME}")"
+if [ "$(echo "${output}" | head -1)" = "${CERTHEADER}" ]; then
+    error "CA already has signed certificate for '${CERTNAME}'"
+fi
+
+### Generate keys and CSR for this host
+[ -s "${PRIVKEYFILE}" ] && error "private key '${PRIVKEYFILE}' already exists"
+[ -s "${PUBKEYFILE}" ] && error "public key '${PUBKEYFILE}' already exists"
+[ -s "${CSRFILE}" ] && error "certificate request '${CSRFILE}' already exists"
+openssl genrsa -out "${PRIVKEYFILE}" 4096
+openssl rsa -in "${PRIVKEYFILE}" -pubout -out "${PUBKEYFILE}"
+openssl req -new -key "${PRIVKEYFILE}" -out "${CSRFILE}" -subj "${CERTSUBJECT}"
+
+### Submit CSR
+curl ${CURLFLAGS} -X PUT -H "Content-Type: text/plain" \
+    --data-binary "@${CSRFILE}" "${CA}/certificate_request/${CERTNAME}"
+
+### Retrieve signed certificate; wait and try again with a timeout
+sleeptime=10
+timewaited=0
+cert="$(curl ${CURLFLAGS} "${CA}/certificate/${CERTNAME}")"
+while [ "$(echo "${cert}" | head -1)" != "${CERTHEADER}" ]; do
+    [ ${timewaited} -ge ${WAITFORCERT} ] && \
+        error "timed-out waiting for certificate to be signed"
+    msg "Waiting for certificate to be signed..."
+    sleep ${sleeptime}
+    timewaited=$((${timewaited}+${sleeptime}))
+    cert="$(curl ${CURLFLAGS} "${CA}/certificate/${CERTNAME}")"
+done
+echo "${cert}" > "${CERTFILE}"
+
+### Verify we got a signed certificate
+if [ -f "${CERTFILE}" ] && [ "$(head -1 "${CERTFILE}")" = "${CERTHEADER}" ]; then
+    if openssl x509 -subject -issuer -noout -in "${CERTFILE}"; then
+        msg "Successfully signed certificate '${CERTFILE}'"
+    else
+        error "invalid signed certificate '${CERTFILE}'"
```

Comment:
A question perhaps for a future update... should we consider blowing away intermediate state? Or is the expectation that if something went wrong, we just fire up a new container?

Reply:
What intermediate state are you referring to?

If something goes wrong, it's either server-side or client-side, and there's nothing we can do if it's server-side.  If it's client-side, and recoverable, I think this script should attempt that.

Not sure what form that would take, but I expect we'll improve this script as we encounter more scenarios like that.